spectrum-offchain-lm/src/data/pool.rs

```diff
@@ -0,0 +1,583 @@
+    pub program_start: u32,
+    pub redeem_blocks_delta: u32,
+    pub program_budget: TypedAssetAmount<Reward>,
+}
+
+impl From<ProgramConfig> for Vec<i32> {
+    fn from(c: ProgramConfig) -> Self {
+        vec![
+            c.epoch_len as i32,
+            c.epoch_num as i32,
+            c.program_start as i32,
+            c.redeem_blocks_delta as i32,
+        ]
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum PoolOperationError {
+    Permanent(PermanentError),
+    Temporal(TemporalError),
+}
+
+#[derive(Debug, Display)]
+pub enum PermanentError {
+    LiqudityMismatch,
+    ProgramExhausted,
+    OrderPoisoned,
+    LowExFee,
+}
+
+#[derive(Debug, Display)]
+pub enum TemporalError {
+    LiquidityMoveBlocked,
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Pool {
+    pub pool_id: PoolId,
+    pub state_id: PoolStateId,
+    pub budget_rem: TypedAssetAmount<Reward>,
+    pub reserves_lq: TypedAssetAmount<Lq>,
+    pub reserves_vlq: TypedAssetAmount<VirtLq>,
+    pub reserves_tmp: TypedAssetAmount<Tmp>,
+    pub epoch_ix: Option<u32>,
+    pub max_error: u64,
+    pub conf: ProgramConfig,
+    pub erg_value: NanoErg,
+}
+
+impl Pool {
+    pub fn pool_nft(&self) -> TypedAssetAmount<PoolNft> {
+        TypedAssetAmount::new(self.pool_id.into(), MAX_VALUE)
+    }
+
+    /// Apply deposit operation to the pool.
+    /// Returns pool state after deposit, new staking bundle and user output in the case of success.
+    /// Returns `PoolOperationError` otherwise.
+    pub fn apply_deposit(
+        self,
+        deposit: Deposit,
+        ctx: ExecutionContext,
+    ) -> Result<(Pool, StakingBundleProto, DepositOutput, ExecutorOutput), PoolOperationError> {
+        if self.num_epochs_remain(ctx.height) < 1 {
+            return Err(PoolOperationError::Permanent(PermanentError::ProgramExhausted));
+        }
+        if !self.epoch_completed() {
+            return Err(PoolOperationError::Temporal(TemporalError::LiquidityMoveBlocked));
+        }
+        let release_vlq = deposit.lq.coerce::<VirtLq>();
+        let release_tmp_amt = self.num_epochs_remain(ctx.height) as u64 * release_vlq.amount;
+        let release_tmp = TypedAssetAmount::new(self.reserves_tmp.token_id, release_tmp_amt);
+        let mut next_pool = self;
+        next_pool.reserves_lq = next_pool.reserves_lq + deposit.lq;
+        next_pool.reserves_vlq = next_pool.reserves_vlq - release_vlq;
+        next_pool.reserves_tmp = next_pool.reserves_tmp - release_tmp;
+        let bundle_key = TypedAssetAmount::new(ctx.mintable_token_id, MAX_VALUE);
+        let bundle = StakingBundleProto {
+            bundle_key_id: bundle_key.to_asset(),
+            pool_id: next_pool.pool_id,
+            vlq: release_vlq,
+            tmp: release_tmp,
+            redeemer_prop: deposit.redeemer_prop.clone(),
+            erg_value: NanoErg::from(BoxValue::SAFE_USER_MIN),
+        };
+        let user_output = DepositOutput {
+            bundle_key,
```

Comment:
So `user_output` will have MAX_VALUE of BundleKey tokens, but doesn't this conflict with this line of the contract: https://github.com/spectrum-finance/ergo-dex/blob/ad099cc64fb35b1c7a9a72d97631a8776281ed75/contracts/lqmining/simple/Deposit.sc#L48?